### ISSUE TITLE
Default to high resolution random number generators

### ISSUE THREAD

**Issue by mainegra** [nrc-cnrc/EGSnrc#594]
There are situations were `ranmar` gives wrong results. It is a better approach to have someone switch to the low resolution ranmar rng intentionally, than having users obtaining wrong results unaware of this issue. Make ranlux (Mortran apps) or the high-resolution version (C++ apps) the default random number generator in EGSnrc. 
_This is especially important as more simulations move to low energies and small scales._

**Comment by ftessier** [comment]
(see discussion also in #292)

**Comment by mainegra** [comment]
**Hi-res vs low-res ranmar rng timing for isotropic photon source:**

- Kerma in 5 cm radius air sphere @ 1m 40 keV  (egs_kerma): 8% slowdown
- DD in 5 cm radius water sphere for 5 keV (egs_app), 
    - ph only: 4% slowdown
    - ph+e- : 8% slowdown
- DD in 15 cm radius water sphere for 200 keV (egs_app), 
    - ph+e- : 7.24% slowdown
- DD in 15 cm radius water sphere for 2 MeV (egs_app), 
    - ph+e- : 7.5% slowdown

In low Z materials the slowdown is about 8% from keV energies up to 2 MeV. In high Z materials the penalty might be higher.

**Comment by MartinMartinov** [comment]
I don't still have the hard numbers, but in my gold nanoparticle simulations when I first started using the high resolution option years ago, I found slowdowns ranging 10-14%

**Comment by ftessier** [comment]
Hmm, that's not negligible. I wonder if we make the low-res version the default, with a prominent _notice_ to turn the hi-res option when "needed" (with a clearer definition), or default to hi-res with a prominent _notice_ stating that one could gain ~10% with the low-res option, at your own risk! The latter is the safer. Will look at ranlux timing, and of course the real answer is `xoshiro256+` 😉

**Comment by mainegra** [comment]
@MartinMartinov thanks for sharing that info! I'm running a case for a Pb sphere and will update above.
@ftessier a proper definition on when to use the hires rng might be harder to find. Perhpaps example cases such as those encountered by Martin, Rowan and Pat. And now the case of low energy photons I have encountered. Seems like we could use a co-op student to work on this in the Summer .... 😉

**Comment by mainegra** [comment]
@ftessier if you are interested here is a bit of a soap opera I discover last night in the world of RNGs researchers:

https://www.pcg-random.org/posts/on-vignas-pcg-critique.html

Spoiler alert: They might be arguing about minutiae in our MC world!

**Comment by ftessier** [comment]
Wow, thanks for that perspective; perhaps `pcg64` is the better contender. Will have to do some more reading on this entertaining drama! Then again, if the code is in the public domain, we can include "all of them" and let the users choose... In terms of hires/lowres, I'd side with making hires the default with some clear messaging about the potential ~10% saving because indeed, it is hard to define broadly in which cases hires is needed.

**Comment by rtownson** [comment]
I agree with highres default with instructions to switch to lowres when the time savings are needed. I think the general philosophy in EGSnrc is that the default settings are accurate in the vast majority of conditions.

**Comment by mainegra** [comment]
@ftessier see above the result ranlux vs ranmar in Mortran. It is actually getting worse!

**Comment by ftessier** [comment]
Darn! Skip all this and implement xoshiro256+ right away? 🤓

**Comment by ftessier** [comment]
Posted in the thread again. Here are my replies regarding the pcg vs xoshiro debate:

https://github.com/nrc-cnrc/EGSnrc/pull/292#issuecomment-641583572

**Comment by ftessier** [comment]
On pcg vs xoshiro: after additional reading, I side with the `xoshiro` generator, for its dead simple code, its speed, and its ability to jump ahead to generate independent random sequences for parallel simulations. In terms of randomness _quality_, I defer to the mathematicians; to me it suffices that it passes the TestU01 BigCrush test.

**Comment by ftessier** [comment]
Only thing left as an implementation detail is how to sample doubles uniformly on [0,1). Do we just take the high bits from the xoshiro256+ 64-bit integer result and plug them into the mantissa of a double? Is that correct? Other suggestions welcome.

**Comment by ftessier** [comment]
I guess using the 53 most significant bits suffices for our purpose in EGSnrc:
```
double random_real_53(void) {
	return ldexp((double)(random64() & ((1ULL << 53) - 1)), -53);
}
```
This is from http://mumble.net/~campbell/tmp/random_real.c, which also proposes a solution for the issue that this is **not** uniform: values in (0, 2e-53) are never picked, for example.

**Comment by mxxo** [comment]
Perhaps it's worth mentioning [ranlux++](https://arxiv.org/abs/1705.03123) as another potential choice? Greatly improved performance over plain ranlux but still "[suitable for use in the most demanding Monte Carlo calculations](https://link.springer.com/article/10.1007/s41781-019-0034-3)". There exists a portable C++ implementation that was recently added to ROOT: https://arxiv.org/abs/2106.02504

**Comment by ftessier** [comment]
I have not looked at the `ranlux` prng recently, because they are (were) painfully slow generators. But an implementation is useful to eventually check other generators. Thanks for the reference, which is discussing some interesting optimizations, so I will look into this again. I quickly note in Figure 2 performance above about 10 ns/number (with conversion), whereas xoshiro is sub 1 ns/number according to https://prng.di.unimi.it.
